File: packages/shared/src/prebake.ts
/**
 * Pre-baked face windows.
 *
 * Generating a face on demand on Modal pays a GPU cold-start (~$). To keep the
 * carousel cheap we pre-render a fixed grid of slot ranges into a Modal Volume
 * once, then serve them as static JPEGs (CPU + Cloudflare cache) for free.
 *
 * Layout: PREBAKE_WINDOW_COUNT windows of PREBAKE_WINDOW_SIZE consecutive
 * slots each, evenly spaced across the slot range and centered within their
 * slice. A new visitor is dropped into a randomly-picked window so most
 * impressions never touch the GPU; the user can still scroll out of the
 * window into "empty" slots which simply show the silhouette fallback.
 *
 * Anchor positions are derived from POPULATION_BASELINE (the static day-zero
 * total), not currentTotalSlots(), so the on-disk file names stay valid as
 * the population number ticks up daily.
 */

// Mirrored from index.ts intentionally to avoid an import cycle. Update both
// in lockstep — the Python side in services/face-generator/prebake.py also
// depends on this constant matching.
const POPULATION_BASELINE = 8_289_468_500n;

export const PREBAKE_WINDOW_COUNT = 100;
export const PREBAKE_WINDOW_SIZE = 100;

const STEP: bigint = POPULATION_BASELINE / BigInt(PREBAKE_WINDOW_COUNT);
const HALF_WINDOW: bigint = BigInt(Math.floor(PREBAKE_WINDOW_SIZE / 2));
const FULL_WINDOW: bigint = BigInt(PREBAKE_WINDOW_SIZE);

/** Start slot of every pre-baked window, in order. */
export function prebakeAnchors(): bigint[] {
  const anchors: bigint[] = [];
  for (let i = 0; i < PREBAKE_WINDOW_COUNT; i++) {
    const center = STEP * BigInt(i) + STEP / 2n;
    const start = center > HALF_WINDOW ? center - HALF_WINDOW : 0n;
    anchors.push(start);
  }
  return anchors;
}

/** Pick one anchor uniformly at random. Used by SSR to land each visitor
 *  inside a fresh window per request. */
export function randomPrebakeAnchor(rng: () => number = Math.random): bigint {
  const idx = Math.floor(rng() * PREBAKE_WINDOW_COUNT);
  return prebakeAnchors()[idx]!;
}

/** O(1) check — does `slot` fall in any pre-baked window? */
export function isPrebaked(slot: bigint): boolean {
  if (slot < 0n || slot >= POPULATION_BASELINE) return false;
  const sliceIdx = slot / STEP;
  if (sliceIdx >= BigInt(PREBAKE_WINDOW_COUNT)) return false;
  const center = STEP * sliceIdx + STEP / 2n;
  const start = center > HALF_WINDOW ? center - HALF_WINDOW : 0n;
  const end = start + FULL_WINDOW;
  return slot >= start && slot < end;
}

/** Pick a random slot inside a specific window. */
export function randomSlotInWindow(anchor: bigint, rng: () => number = Math.random): bigint {
  const offset = BigInt(Math.floor(rng() * PREBAKE_WINDOW_SIZE));
  return anchor + offset;
}

File: services/face-generator/prebake.py
"""Pre-baked face windows.

Mirrors packages/shared/src/prebake.ts — keep the constants in lockstep when
either side changes. Anchor positions are derived from POPULATION_BASELINE
(static day-zero total) so the on-disk filenames in the prebake volume stay
valid as the daily total ticks up.
"""

from __future__ import annotations

POPULATION_BASELINE = 8_289_468_500
PREBAKE_WINDOW_COUNT = 100
PREBAKE_WINDOW_SIZE = 100

_STEP = POPULATION_BASELINE // PREBAKE_WINDOW_COUNT
_HALF = PREBAKE_WINDOW_SIZE // 2


def prebake_anchors() -> list[int]:
    """Start slot of every pre-baked window, in order."""
    anchors: list[int] = []
    for i in range(PREBAKE_WINDOW_COUNT):
        center = _STEP * i + _STEP // 2
        start = center - _HALF if center > _HALF else 0
        anchors.append(start)
    return anchors


def is_prebaked(slot: int) -> bool:
    """O(1) check — does ``slot`` fall in any pre-baked window?"""
    if slot < 0 or slot >= POPULATION_BASELINE:
        return False
    slice_idx = slot // _STEP
    if slice_idx >= PREBAKE_WINDOW_COUNT:
        return False
    center = _STEP * slice_idx + _STEP // 2
    start = center - _HALF if center > _HALF else 0
    end = start + PREBAKE_WINDOW_SIZE
    return start <= slot < end
